pmd-core/src/test/java/net/sourceforge/pmd/lang/ast/AbstractNodeTest.java

```diff
@@ -0,0 +1,233 @@
+        assertNull(rootNode.jjtGetParent());
+        for (final Node aChild : children) {
+            assertNull(aChild.jjtGetParent());
+        }
+    }
+
+    /**
+     * Explicitly tests the {@code remove} method, and implicitly the {@code removeChildAtIndex} method.
+     * These are border cases as grandchildren nodes do not have any child.
+     */
+    @Test
+    @Parameters(method = "childrenAndGrandChildrenIndexes")
+    public void testRemoveGrandChildNode(final Integer childIndex, final Integer grandChildIndex) {
+        final Node child = rootNode.jjtGetChild(childIndex);
+        final Node grandChild = child.jjtGetChild(grandChildIndex);
+
+        // Check that the child has the expected properties
+        assertEquals(NUM_GRAND_CHILDREN, child.jjtGetNumChildren());
+        assertEquals(0, grandChild.jjtGetNumChildren());
+        assertEquals(child, grandChild.jjtGetParent());
+
+        // Do the actual removal
+        grandChild.remove();
+
+        // Check that conditions have been successfully changed
+        assertEquals(NUM_GRAND_CHILDREN - 1, child.jjtGetNumChildren());
+        assertEquals(0, grandChild.jjtGetNumChildren());
+        assertNull(grandChild.jjtGetParent());
+    }
+
+    /**
+     * Explicitly tests the {@code removeChildAtIndex} method.
+     */
+    @Test
+    @Parameters(method = "childrenIndexes")
+    public void testRemoveRootNodeChildAtIndex(final Integer childIndex) {
+        final Node[] originalChildren = new Node[rootNode.jjtGetNumChildren()];
+
+        // Check that prior conditions are OK
+        for (int i = 0; i < originalChildren.length; i++) {
+            originalChildren[i] = rootNode.jjtGetChild(i);
+            assertEquals(i, originalChildren[i].jjtGetChildIndex());
+            if (i > 0) {
+                assertNotEquals(originalChildren[i - 1], originalChildren[i]);
+            }
+        }
+        assertEquals(NUM_CHILDREN, rootNode.jjtGetNumChildren());
+
+        // Do the actual removal
+        rootNode.removeChildAtIndex(childIndex);
+
+        // Check that conditions have been successfully changed
+        assertEquals(NUM_CHILDREN - 1, rootNode.jjtGetNumChildren());
+        int j = 0;
+        for (int i = 0; i < rootNode.jjtGetNumChildren(); i++) {
+            if (j == childIndex) { // Skip the removed child
+                j++;
+            }
+            // Check that the nodes have been rightly shifted
+            assertEquals(originalChildren[j], rootNode.jjtGetChild(i));
+            // Check that the child index has been updated
+            assertEquals(i, rootNode.jjtGetChild(i).jjtGetChildIndex());
+            j++;
+        }
+    }
+
+    /**
+     * Explicitly tests the {@code removeChildAtIndex} method.
+     * Test how invalid indexes cases are handled.
+     */
+    @Test
+    public void testRemoveChildAtIndexWithInvalidIndex() {
+        // No assert as the test is considered passed if no exception is thrown
```

Comment:
to make this explicit, you should `try-catch` the whole block, and add a `fail` on the catch. That would also comply with our [JUnitTestsShouldIncludeAssert](https://pmd.github.io/pmd/pmd_rules_java_junit.html#junittestsshouldincludeassert) rule